pkg/cache/cache.go

```diff
@@ -0,0 +1,283 @@
+	Prime(key string, value interface{})
+
+	// Deletes the value identified by the given key from the cache, and
+	// generates an update on the queue if a value was deleted.
+	Delete(key string)
+
+	// Lists the keys currently in the cache.
+	ListKeys() []string
+
+	// Returns true when the cache has synced with the datastore,
+	// false until that point.
+	HasSynced() bool
+
+	// Starts the cache.
+	Run(stopChan chan struct{})
+
+	// Sets kubernetes cache indexer
+	SetIndexer(indexer k8sCache.Indexer)
+
+	// Get workqueue
+	GetQueue() workqueue.RateLimitingInterface
+}
+
+// ResourceCacheArgs struct passed to constructor of ResourceCache.
+// Groups togather all the arguments to pass in single struct.
+type ResourceCacheArgs struct {
+	// ListFunc returns a list of objects.  Responsible for filtering any
+	// objects which should not be monitored by the cache / controller.
+	ListFunc func() ([]interface{}, error)
+
+	// Takes an object and returns the key string which identifies it in the cache.
+	// e.g. namespace.name
+	KeyFunc func(obj interface{}) string
+
+	// The channel on which modified keys will be sent, if given.
+	OutChan chan string
+
+	// Calico Client
+	Client *calicoClient.Client
+
+	// Type of object cache will hold
+	// Set() API will verfiy the object type before storing it in cache
+	ObjectType string
+}
+
+// CalicoCache implements ResourceCache interface
+// Cache only stores pointer to objects instead of actual objects
+type calicoCache struct {
+	threadSafeCache *cache.Cache                    // Underlaying threadsafe implementation of cache
+	workqueue       workqueue.RateLimitingInterface // Workqueue
+	calicoClient    *calicoClient.Client            // Clinet to Calico ETCD datastore
+	synced          bool                            // Flag set to true when cache is synced with etcd datastore
+	k8sCacheIndexer k8sCache.Indexer                // K8S cache indexer used while priming of calicoCache.
+	ListFunc        func() ([]interface{}, error)   // Function that returns a list of objects.
+	KeyFunc         func(obj interface{}) string    // Function that returns key string which identifies it in the cache.
+	ObjectType      string                          // Type of object cache will hold
+}
+
+// NewResourceCache Constructor for ResourceCache.
+// Requires calico client to prime the cache
+// Cache only allows adding objects of type `objectType`
+func NewResourceCache(args ResourceCacheArgs) ResourceCache {
+
+	return &calicoCache{
+		threadSafeCache: cache.New(cache.NoExpiration, cache.DefaultExpiration),
+		workqueue:       workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		calicoClient:    args.Client,
+		synced:          false,
+		ListFunc:        args.ListFunc,
+		KeyFunc:         args.KeyFunc,
+		ObjectType:      args.ObjectType,
+	}
+}
+
+func (c *calicoCache) Set(key string, newObj interface{}) {
+
+	if reflect.TypeOf(newObj).String() == c.ObjectType {
```

Comment:
I think if we're going to check this, it would be good to do it as a separate if statement.  I also think we should exit (this indicates a bug in the code, not normal operation or even an error scenario)

e.g.

```
if reflect.TypeOf() ... {
  glog.Fatal("")
}

// See if it exists in the cache. 
...
```